### PR TITLE
Use hash table to store AP information

### PR DESCRIPTION
**Pull request by dungru**
1. Use `strspn` and `strcspn` to parsing the parameter.
2. Store the AP information into hash table which is `ap_info_hash`

### CODE REVIEW

**Comment by jserv** [comment]
I would invite @lekoOwO for reviewing.

**Comment by lekoOwO** [comment]
What's the pro of storing AP info?

The main purpose of generating bssid from `hash(ssid)` is to make it stateless, so we do not need to store it.
If we're going to store it into a hash table, the bssid can be generated from a random number instead of `hash(ssid)`.

**Comment by ldotrg** [comment]
> What's the pro of storing AP info?
> 
> The main purpose of generating bssid from `hash(ssid)` is to make it stateless, so we do not need to store it. If we're going to store it into a hash table, the bssid can be generated from a random number instead of `hash(ssid)`.

The table is not

**Comment by ldotrg** [comment]
> > > What's the pro of storing AP info?
> > > The main purpose of generating bssid from `hash(ssid)` is to make it stateless, so we do not need to store it. If we're going to store it into a hash table, the bssid can be generated from a random number instead of `hash(ssid)`.
> > 
> > 
> > Actually, if you don't change SSID, the murmurhash result will always same. The AP info always get update when `module_param` has changed and trigger STA scan. The hash table provides a more efficient way to get AP info with O(1). On the other hand, this table could extend other AP information fields (i.e. security type) not just store the bssid. The original method iterate every SSID string, it cost O(n).
> 
> I see.
> 
> Then we'd need a limit for stored AP Infos. Also the bssid can now be generated randomly instead of hashing ssid.
> 
> It might help understanding the code if we comment that the field in AP info is extendable.

If we generate the BSSID such as `eth_random_addr()` in every scan stage, how could we make sure same SSID will not mapping to different BSSID. When BSSID is changed during STA is under following probe/associate/4-way hand shake/port secured staged or even AP and STA are connected. The WPA supplicant will found when  the SSID and BSSID is not match and abort the connection.

Correct me if I am wrong.

File: vwifi.c [hunk 1]
```diff
@@ -44,10 +48,20 @@ struct owl_ndev_priv_context {
...
+    u8 bssid[ETH_ALEN];
+    char ssid[SSID_MAX_LENGTH];
+};
+
+static char *ssid_list = DEFAULT_SSID_LIST "[MyHomeWiFi01]";
 module_param(ssid_list, charp, 0644);
 MODULE_PARM_DESC(ssid_list, "Self-defined SSIDs.");
-static const char sdssid_delim[] = "]";
+/* AP Database */
+DECLARE_HASHTABLE(ess_info_tbl, 4);
```
Comment (jserv) [inline vwifi.c:64]: How about renaming it to `ssid_table`?

File: vwifi.c [hunk 2]
```diff
@@ -72,46 +86,90 @@ static s32 rand_int(s32 low, s32 up)
...
+        hash_for_each_possible_safe (ess_info_tbl, ap, tmp, node, key) {
+            if (strncmp(token, ap->ssid, n) == 0) {
+                ssid_exist = true;
+                break;
+            }
+        }
+        if (ssid_exist)  // SSID exist
+            continue;
+        ap = kzalloc(sizeof(struct ap_info_entry_t), GFP_KERNEL);
+        if (ap == NULL) {
```
Comment (jserv) [inline vwifi.c:146]: Change to `if (!ap) {`

File: vwifi.c [hunk 3]
```diff
@@ -44,10 +48,20 @@ struct owl_ndev_priv_context {
     struct wireless_dev wdev;
 };
 
-static char *ssid_list = "";
+/*
+ * AP information table entry. Entry field is extendable.
```
Comment (jserv) [inline vwifi.c:52]: I did not get the meaning of "Entry field is extendable." What did you tend to express?

Comment (dungru) [reply vwifi.c:52]: Store other AP information such as Capability information which is from the beacon frame.
I could remove the comment "Entry field is extendable."

**Pull request merged**